src/ide/code_actions/fill_trait_members.rs

```diff
@@ -0,0 +1,285 @@
+/// Returns None if function has a default implementation.
+fn function_code(
+    db: &AnalysisDatabase,
+    id: TraitFunctionId,
+    type_concretizer: &mut SubstitutionRewriter<'_>,
+) -> Option<String> {
+    // Do not complete functions that have default implementations.
+    if db.trait_function_body(id).ok()?.is_some() {
+        return None;
+    }
+
+    let signature = db.trait_function_signature(id).ok()?;
+
+    let generic_parameters = db.trait_function_generic_params(id).ok()?;
+    let generic_parameters_bracket = if generic_parameters.is_empty() {
+        String::new()
+    } else {
+        let formatted_parameters = generic_parameters
+            .into_iter()
+            .map(|parameter| generic_parameter_code(db, parameter, type_concretizer))
+            .collect::<Option<Vec<_>>>()?
+            .join(", ");
+
+        format!("<{formatted_parameters}>")
+    };
+
+    let parameters = signature
+        .params
+        .iter()
+        .map(|parameter| function_parameter(db, parameter, type_concretizer))
+        .collect::<Option<Vec<_>>>()?
+        .join(", ");
+
+    let name = id.name(db);
+    let title = Some(format!("fn {name}{generic_parameters_bracket}({parameters})"));
+
+    let return_type = type_concretizer.rewrite(signature.return_type).ok()?;
+    let return_type =
+        if return_type.is_unit(db) { None } else { Some(format!("-> {}", return_type.format(db))) };
+
+    let implicits = match &signature.implicits[..] {
+        [] => None,
+        types => Some(format!("implicits({})", types.iter().map(|ty| ty.format(db)).join(", "))),
+    };
+
+    let nopanic = if !signature.panicable { Some(String::from("nopanic")) } else { None };
+
+    let body: Option<String> = Some(String::from("{}"));
+
+    Some([title, return_type, implicits, nopanic, body].into_iter().flatten().join(" "))
+}
+
+/// Formats [`GenericParam`] to be used in function's declaration.
+/// Substitutes generic parameters with proper concrete types given in impl,
+/// keeps free-standing (not belonging to the trait) generics unchanged.
+fn generic_parameter_code(
+    db: &AnalysisDatabase,
+    parameter: GenericParam,
+    type_concretizer: &mut SubstitutionRewriter<'_>,
+) -> Option<String> {
+    match parameter {
+        GenericParam::Const(param) => {
+            Some(format!("const {}: {}", param.id.format(db), param.ty.format(db)))
+        }
+
+        GenericParam::Impl(param) => {
+            let concrete_trait = param.concrete_trait.ok()?;
+            let trait_name = concrete_trait.name(db);
+            let trait_generic_arguments = concrete_trait.generic_args(db);
+
+            let generic_arguments_bracket = if trait_generic_arguments.is_empty() {
+                String::new()
+            } else {
+                let formatted_arguments = trait_generic_arguments
+                    .into_iter()
+                    .map(|argument| generic_argument_code(db, argument, type_concretizer))
```

Comment:
[nit] If all generics were substituted with concrete traits here (e.g. `+Drop<T>` with `+Drop<felt252>`) cannot we just drop the trait contraint? Or does cairo not allow it? It is really a nitpick, you can create an issue for that

Reply:
It's safer to generate this more explicit version probably

Reply:
Yep, it is not allowed in cairo, just checked it. Resolving